### PR TITLE
Get GPU topology from sysfs

### LINKED ISSUE #3559: Derive PlatformTopo information about GPUs from sysfs
> Seems like we should be able to derive the PlatformTopo information about GPUs from this class map.  Maybe we could wrap the map in an accessor class and reuse it in the PlatformTopo imp?

_Originally posted by @cmcantalupo in https://github.com/geopm/geopm/pull/3556#discussion_r1718700340_
            

> [11:22 AM] Cantalupo, Christopher M
> https://github.com/geopm/geopm/blob/dev/libgeopmd/src/GPUTopo.hpp
> geopm/libgeopmd/src/GPUTopo.hpp at dev · geopm/geopm
> Global Extensible Open Power Manager. Contribute to geopm/geopm development by creating an account on GitHub.
>  
> [11:23 AM] Cantalupo, Christopher M
> i think we just need to wrap it in a GPUTopo?

### PR DESCRIPTION
**Pull request by dannosliwcd**
- Fixes #3559
- Depends on #3561

### CODE REVIEW

File: libgeopmd/src/DrmGpuTopo.cpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    auto driver_with_max_cards_it = std::max_element(
+        driver_card_paths.begin(), driver_card_paths.end(),
+        [](const DriverCards &lhs, const DriverCards &rhs) {
+            return lhs.second.size() < rhs.second.size();
+        });
+    if (driver_with_max_cards_it == driver_card_paths.end()) {
+        // This should only happen if driver_card_paths (and all_cards) are empty.
+        throw std::logic_error(std::string(__func__) + ": No max-count is defined for driver card paths");
+    }
+    return *driver_with_max_cards_it;
```
Comment (cmcantalupo) [inline libgeopmd/src/DrmGpuTopo.cpp:112]: I think this is returns the driver that has the largest number of cards associated with it.  Maybe a comment about the return value could clarify this.

File: libgeopmd/src/DrmGpuTopo.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                tiles_per_card = tile_paths_in_card.size();
+            }
+            else {
+                if (tiles_per_card != static_cast<int>(tile_paths_in_card.size())) {
+                    throw Exception("DrmGpuTopo::" + std::string(__func__) +
+                                        ": Mixed counts of gpu_chip per gpu are not "
+                                        "supported. Encountered at least one gpu with " +
+                                        std::to_string(tiles_per_card) + " tiles per card and " +
+                                        std::to_string(tile_paths_in_card.size()) +
+                                        " tiles on " + card_path,
```
Comment (cmcantalupo) [inline libgeopmd/src/DrmGpuTopo.cpp]: might be easier to read with an std::ostringstream

File: libgeopmd/src/Helper.cpp [hunk 3]
```diff
@@ -401,4 +401,31 @@ namespace geopm
...
+        if (err) {
+            throw Exception("read_symlink_target(): failed to lstat " + symlink_path,
+                            errno ? errno : GEOPM_ERROR_RUNTIME,
+                            __FILE__, __LINE__);
+        }
+
+
+        size_t symlink_strlen = stat_buf.st_size > 0
+            ? stat_buf.st_size
+            : NAME_MAX; // May happen for symlinks in sysfs
```
Comment (cmcantalupo) [inline libgeopmd/src/Helper.cpp]: PATH_MAX?

**Review by cmcantalupo — COMMENTED** [review]
Nice work, and done so quickly!  Just a few comments about minor things.

**Review by dannosliwcd — COMMENTED** [review]
rebase and squash fixups before merging

**Review by cmcantalupo — APPROVED** [review]

**Pull request merged**